### DISCUSSION TITLE
Support calling args as object instead of tuples for contract functions

### DISCUSSION THREAD

**Discussion by shirotech** [wevm/viem#2534] [Idea / Feature Request]
When supplying the args for calling a contract function, it is currently in arrays or strongly typed tuples. However, it is kinda hard to know the order and know which index is referring to which param. By using named object props mapping to the Abi function order, it makes code alot more readable. I've got a little helper below that converts the object to the typed tuple and finally pass it to the function call. However it's kinda repetitive, not having the ability to call it directly from a contract object.

Would be nice to do something like this:

```ts
tokenContract.read.balanceOf({account: "0xabc"});
```

rather than below workaround:

```ts
tokenContract.read.balanceOf(mapArgs(erc20Abi, "balanceOf", { account: "0xabc" }));
tokenContract.read.balanceOf(["0xabc"]); // result
```

and below is snippet of `mapArgs`

```ts
import type {
  Abi,
  AbiParametersToPrimitiveTypes,
  AbiParameterToPrimitiveType,
  ExtractAbiFunction,
} from "abitype";
import { type ContractFunctionName } from "viem";

export function mapArgs<
  abi extends Abi,
  functionName extends ContractFunctionName<abi>,
  abiFunction extends ExtractAbiFunction<abi, functionName>,
  args extends {
    [K in abiFunction["inputs"][number] as NonNullable<K["name"]>]: AbiParameterToPrimitiveType<
      K,
      "inputs"
    >;
  },
>(
  abi: abi,
  functionName: functionName,
  args: args,
): AbiParametersToPrimitiveTypes<abiFunction["inputs"]> {
  return abi
    .find((item): item is abiFunction => item.type === "function" && item.name === functionName)!
    .inputs.map((input) => args[input.name as keyof args]) as any;
}

```

A single param is pretty straightforward, but when you have 3 or more params for a contract input, such as swapping a token, it can get confusing at times especially dealing with different DEX that has their own signatures.

Please note that I'm not asking if we can have a breaking change to replace existing signature, but also support it alongside of tuples.

Thank you for your considerations, and hope this can be implemented.

**Comment by shirotech** [comment]
Hi @jxom was wondering if you have any thoughts on this and is it something that can be supported? Thank you 🙏

**Comment by Yuripetusko** [comment]
FYI this was discussed in discord recently (but no final conclusion yet on the approach)

https://discord.com/channels/1156791276818157609/1268170772036325386

**Comment by Yuripetusko** [comment]
Really cool, thank you for this! I used this to make a reversed version too. That for given function output array returns a named object

```
export function mapFunctionOutputToNamedObject<
  abi extends Abi,
  functionName extends ContractFunctionName<abi>,
  abiFunction extends ExtractAbiFunction<abi, functionName>,
  args extends AbiParametersToPrimitiveTypes<abiFunction['outputs']>,
>(
  abi: abi,
  functionName: functionName,
  args: args,
): {
  [K in abiFunction['outputs'][number] as NonNullable<K['name']>]: AbiParameterToPrimitiveType<
    K,
    'outputs'
  >;
} {
  const output = abi
    .find((item): item is abiFunction => item.type === 'function' && item.name === functionName)!
    .outputs.reduce((acc, output, index) => {
      acc[(output.name as keyof args) || index] = args[index];
      return acc;
    }, {} as any);
  return output;
}
```

So here I have this

```
[
    "1000000000000041127",
    "0xfbea1b97406C6945D07F50F588e54144ea8B684f",
    true,
    false,
    false,
    true,
    "32",
    "0",
    "0",
    "0"
]
```

And the util gives me this back:

```
{
    "currentSettlementId": "1000000000000041127",
    "owner": "0xfbea1b97406C6945D07F50F588e54144ea8B684f",
    "nftIsBusy": true,
    "nftIsTravelling": false,
    "nftIsSurveyingSettlement": false,
    "nftIsInActiveQuest": true,
    "currentPlayerQuestId": "32",
    "travelEndTime": "0",
    "surveyEndTime": "0",
    "questEndTime": "0"
}
```

**Comment by shirotech** [comment]
I've written a gist that takes advantage of `client.extend()` to elegantly support calling and returning objects instead of tuples. https://gist.github.com/shirotech/5d934f0d827b79fe9ab74f993d0825a3

Please star if you like it :)
